test/net/sourceforge/kolmafia/session/QuestManagerTest.java

```diff
@@ -3550,4 +3550,54 @@ public void doNotProgressQuestPyramidOnUhOh() {
       }
     }
   }
+
+  @Nested
+  class WizardOfEgo {
+    @ParameterizedTest
+    @CsvSource({"staring_into_nothing, 4", "into_the_maw_of_deepness, 5", "take_a_dusty_look, 6"})
+    public void canParseTowerRuinsChoices(String html, int expectedResult) {
+      var cleanups = new Cleanups(withQuestProgress(Quest.EGO, QuestDatabase.UNSTARTED));
+
+      try (cleanups) {
+        String path = "request/test_adventure_tower_ruins_" + html + ".html";
+        var request = new GenericRequest("adventure.php?snarfblat=22");
+        request.responseText = html(path);
+        QuestManager.handleQuestChange(request);
+        assertThat(Quest.EGO, isStep(expectedResult));
+      }
+    }
+
+    @Test
```

Comment:
OK. I think I will insist that you use our standard "Cleanups" method.

This test could/should be this:

```
     @Test
     public void doNotUnfinishQuest() {
      var builder = new FakeHttpClientBuilder();
      var client = builder.client;
      var cleanups =
          new Cleanups(
              withHttpClientBuilder(builder),
              withQuestProgress(Quest.EGO, QuestDatabase.QuestDatabase.FINISHED));
      try (cleanups) {
        client.addResponse(200, html("request/test_adventure_tower_ruins_staring_into_nothing.html"));
        var request = new GenericRequest("adventure.php?snarfblat=22");
        request.run();
        assertThat(Quest.EGO, isStep(QuestDatabase.FINISHED));
      }
    }
```

Note that it:
1) actually submits the request, and verifieds that the expected quest change happens, rather than calling the magic QuestManager method that does that.
2) uses withQuestProgress to set the property and clean up the property after the quest is finished.

This is our standard test convention. Do it for all your tests. :)

Reply:
I really appreciate this feedback! Everything you said makes perfect sense to me and I can see how it's a clear improvement.